include/autoppl/expression/binop.hpp

```diff
@@ -0,0 +1,97 @@
+#pragma once
+
+#include <type_traits>
+#include <autoppl/util/traits.hpp>
+#include <autoppl/expression/var_expr.hpp>
+#include <autoppl/expression/variable.hpp>
+
+namespace ppl {
+
+template <class LHSValueType, class RHSValueType, class BinaryOp>
+struct BinaryOpNode 
+{
+	using binop_result_t = std::common_type_t<
+		typename node_traits<LHSValueType>::binop_result_t,
+		typename node_traits<RHSValueType>::binop_result_t
+			>;
+
+	BinaryOpNode(Variable<LHSValueType> lhs, Variable<RHSValueType> rhs)
+		: lhs_{lhs}, rhs_{rhs}
+	{}
+	
+	binop_result_t get_value() 
+	{
+		LHSValueType lhs_value = lhs_.get_value();
+		RHSValueType rhs_value = rhs_.get_value();
+		return BinaryOp::evaluate(lhs_value, rhs_value);
+	}
+
+private:
+	Variable<LHSValueType> lhs_;
+	Variable<RHSValueType> rhs_;
+
+};
+
+
+template <class LHSValueType, class RHSValueType>
+struct BinaryOp {
+	
+	using binop_result_t = std::common_type_t<
+		typename node_traits<LHSValueType>::binop_result_t,
+		typename node_traits<RHSValueType>::binop_result_t
+			>;
+
+	binop_result_t evaluate();
+
+};
+
+template <class LHSValueType, class RHSValueType>
+struct AddOp : BinaryOp<LHSValueType, RHSValueType> {
+	
+	using binop_result_t = std::common_type_t<
```

Comment:
can replace with `using BinaryOp<LHSValueType, RHSValueType>::binop_result_t;`. You may need to add `typename` in between `using` and the `BinaryOp...`.